### ISSUE TITLE
#selectionBar missing sometimes

### ISSUE THREAD

**Issue by robrez** [PolymerElements/paper-tabs#34]
#selectionBar is "missing" (width, left are both 0 in `_positionBar`) after re-showing an element containing the paper-tabs.

The tabs are hidden conditionally, either via a `dom-if` template OR a `hidden$` attribute. 

For the `hidden` variation of the problem, I can work-around the issue by calling `_onResize` via  async. (See `_doEvil` in my code block below).

Regarding the `dom-if` variation...
I noticed that .5 tabs have some stuff going on in `domReady` that is missing from the 1.0 tabs and another issue has been reported here that sounds related (#31). The solution outlined there (I modified it slightly to use async so as to behave more like domReady in 0.5), does not solve my problem because `attached` does not execute again after the conditional template is once-more truthy... attached executes only once.

Element to duplicate - note that the following is "working" because of the `_doEvil` function:
```
<link rel="import" href="../polymer/polymer.html">
<link rel="import" href="../iron-flex-layout/classes/iron-flex-layout.html">
<link rel="import" href="../paper-tabs/paper-tabs.html">
<link rel="import" href="../paper-tabs/paper-tab.html">
<link rel="import" href="../iron-icon/iron-icon.html">
<link rel="import" href="../iron-icons/iron-icons.html">
<link rel="import" href="../paper-icon-button/paper-icon-button.html">

<dom-module id="tabs-test">
  <style>
    :host {
      display: block;
      position: relative;
      font-family: sans-serif;
      font-size: 14px;
      color: rgba(0,0,0,.87);
    }

    .toolbar {
      min-height: 48px;
      background-color: #00BCD4;
      margin-bottom: 16px;
    }
  </style> 
  <template>
    <div class="toolbar layout horizontal center" hidden$="{{_isPage3(selected)}}">
      <paper-tabs id="tabs" selected="{{selected}}">
        <paper-tab>PAGE 1</paper-tab>
        <paper-tab>PAGE 2</paper-tab>
      </paper-tabs>
      <span class="flex"></span>
      <paper-icon-button icon="create" on-tap="_showPage3"></paper-icon-button>
    </div>

    <div class="toolbar layout horizontal center" hidden$="{{!_isPage3(selected)}}">
       <paper-icon-button icon="arrow-back" on-tap="_showPage1"></paper-icon-button>
       <span>go back</span>
    </div>

    <div><span>Selected page: &nbsp;</span><span>{{selected}}</span></div>
  </template>

<script>
  Polymer({
    is: 'tabs-test',
 
    properties: {
      selected: {
        type: Number,
        value: 0,
        notify: true
      }
    },

    _showPage1: function(e) {
      this.selected = 0;
      this.async(this._doEvil);
    },

    _doEvil: function() {
      if(this.$.tabs) {
          this.$.tabs._onResize();
      }
    },

    _showPage3: function(e) {
      this.selected = 2;
    },

    _isPage3: function(selected) {
      return selected === 2;
    }

  });
</script>

</dom-module>
```

Note that the code above is using the "hidden" approach, not the dom-if.

I originally reported the issue in the wrong repository:
https://github.com/Polymer/paper-tabs/issues/68

**Comment by abourget** [comment]
Seeing same thing here.

**Comment by cdata** [comment]
Thanks, this bug has been confirmed. I will take a look.

**Comment by grokys** [comment]
+1

**Comment by ronnyroeller** [comment]
+1

**Comment by tdesc** [comment]
Paper-tabs inside Paper-toolbar located in 2 pages ( neon-animated-pages ).
After switch between pages #selectionBar loss "transform: translate3d(33.3321%, 0px, 0px) scaleX(0.333321)" it becomes translate3d(0%, 0px, 0px) scaleX(0)

**Comment by robrez** [comment]
@tdesc my workaround for the same problem is to listen for `on-neon-animation-finish` on the pages element. This gets fired when the animation is complete. Do something along the lines of this  in the handler:
```
  _onNeonAnimationFinish: function(e, detail) {
    detail.toPage.fixYourTabs();
  }


  // - elsewhere - 'fix your tabs' is something like this
  //   (similar to the "_doEvil" function in my first example):
  fixYourTabs: function() {
    this.$.myTabs.notifyResize();
  }
```

**Comment by tdesc** [comment]
@robrez Thanks for workaround! At least tabs more usable now. 
Still hope bug will be fixed =)

**Comment by tbone587** [comment]
I am having the same issue.  Pleaseeeeeeeee fix.

**Comment by bicknellr** [comment]
After some discussion about this, the best solution at the moment seems to be to take the same approach as @robrez and trigger `notifyResize` on the tabs (from `IronResizableBehavior`) in response to events relevant to the visibility / layout of your `paper-tabs`. `IronResizableBehavior`, which `paper-tabs` relies on to know when it needs to perform layout updates, requires buy-in from the host using `paper-tabs`. The reason being that there's no single, definitive way (or even a limited number of ways) to know that the visibility / layout of an element has changed without polling. :/

For a `dom-if` template, the relevant event is [`dom-change`](https://github.com/Polymer/polymer/blob/6467ae11c4d4d89ee60d6cba44499fca7e5702d2/src/lib/template/dom-if.html#L114) - [here's an example](http://jsbin.com/koqidi/edit?html,output).

**Comment by robrez** [comment]
After having more time to work w/ tabs under various scenarios, I agree w/ @bicknellr - there are far too many ways to change visibility/layout of tabs to create a one-size-fits-all 'fix'. The parent (or ancestor[s]) should notify the tabs.

I guess we can close the issue? Moving forward, I think some documentation or examples will prove useful.

There are many ways to notify distributed children or effective children [descendant tabs] that a visibility-state change has occured. To help anyone who may encounter problems w/ tabs, here are a handful of contrived techniques that can be used to prevent issues w/ the selection bar. 


#### Tabs inside of a dom-if
Use either a property observer or the `dom-change` event
```
  <template is="dom-if" if="[[someBool]]" on-dom-change="fixTabs">
    <paper-tabs id="tabs">...</paper-tabs>
  </template>
...
  fixTabs: function(e) {
    if (e.target.if) {
      this.$$('#tabs').notifyResize();
    }
  }
```
Depending on the context, you may be able to simply add `restamp` to the dom-if template..
> If the if property becomes falsy again, by default all stamped elements are hidden (but remain in the DOM tree). This provides faster performance should the if property become truthy again. To disable this behavior, set the restamp property to true. This results in slower if switching behavior as the elements are destroyed and re-stamped each time.

```
  <template is="dom-if" if="[[someBool]]" restamp>
    <paper-tabs id="tabs">...</paper-tabs>
  </template>
```


#### Tabs inside of hidden$ conditional attribute
This should similar to dom-if..  One approach is to utilize a property observer that notifies the tabs.
```
  <div hidden$="[[!someBool]]">
    <paper-tabs id="tabs">...</paper-tabs>

    <!-- note: 'observed-prop' could have an observer that notifies tabs -->
    <my-element-with-tabs id="myElementWithTabs" observed-prop="[[someBool]]">
    </my-element-with-tabs>
  </div>
...
  properties: {
    someBool: {
      type: Boolean,
      observer: 'onSomeBoolChanged'
    }
  },

  onSomeBoolChanged: function(someBool, oldValue) {
    if (someBool) {
      this.fixTabs();
    }
  },

  fixTabs: function(e) {
      this.$.tabs.notifyResize();

      // this could be a function call or a property
      this.$.myElementWithTabs.fixTabs();
  }
```


#### Tabs inside of neon-animated-pages
Use `neon-animation-finish` event
```
  <neon-animated-pages
        on-neon-animation-finish="fixTabs"
        attr-for-selected="data-route"
        selected="{{route}}">

    <div data-route="p1">Page 1</div>

    <div data-route="p2">
      <h2>Page 2</h2>
      <paper-tabs id="tabs"></paper-tabs>
    </div>

    <my-element-that-has-tabs data-route="p3">
    </my-element-that-has-tabs>
  <neon-animated-pages>
...
  fixTabs: function(e, detail) {
    var route = detail.toPage.getAttribute("data-route") || this.route;
    if (route == 'p2') {
      //maybe tabs are in your light dom
      this.$$('#tabs').notifyResize();
    } else if (detail.toPage.notifyResize) {
      //maybe tabs are in another element that provides a mechanism for notifying tabs
      // this could be a function call or a property
      detail.toPage.notifyResize();  //detail.toPage.isVisible = true;
    }
  }
```


#### Tabs inside of iron-pages
Use `iron-select` event
```
  <iron-pages
        on-iron-select="fixTabs"
        attr-for-selected="data-route"
        selected="{{route}}">

    <div data-route="p1">Page 1</div>

    <div data-route="p2">
      <h2>Page 2</h2>
      <paper-tabs id="tabs"></paper-tabs>
    </div>

    <my-element-that-has-tabs data-route="p3">
    </my-element-that-has-tabs>
  <iron-pages>
...
  fixTabs: function(e, detail) {
    var route = detail.item.getAttribute("data-route") || this.route;
    if (route == 'p2') {
      //maybe tabs are in your light dom
      this.$$('#tabs').notifyResize();
    } else if (detail.item.notifyResize) {
      //maybe tabs are in another element that provides a mechanism for notifying tabs
      // this could be a function call or a property
      detail.item.notifyResize();  //detail.item.isVisible = true;
    }
  }
```

**Comment by bicknellr** [comment]
Thanks for the detailed examples, @robrez! These will be great for anyone else running into this issue.

**Comment by DanielJoyce** [comment]
Yeah this needs to be documented. Its biting me too.
